fixes biobag reproductive extract interaction bug

MNarath1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This PR fixes a bug that would not allow you to feed monkey cubes to the reproductive extract directly with the biobag.
I am not completly sure how this broke trough the only thing that changed in the call chain for this interaction in the last few months was from https://github.com/BeeStation/BeeStation-Hornet/pull/3883 and i guess https://github.com/BeeStation/BeeStation-Hornet/pull/4623 too trough thats purly specuation at this point.
closes: https://github.com/BeeStation/BeeStation-Hornet/issues/4566
closes: https://github.com/BeeStation/BeeStation-Hornet/issues/4701

## Why It's Good For The Game

You can feed monkeycubes to reproductive extracts directly with the biobag again

## Changelog
:cl:
fix: fixes biobag not allowing you to feed reproductive extract directly
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MNarath1 committed:
fixes biobag repro interaction

Not an ideal fix but eh

RDS88-beeman commented:
FUCKING F I N A L L Y

MNarath1 committed:
consistancy screech

PowerfulBacon commented on code/datums/components/storage/storage.dm:
```diff
@@ -178,7 +178,8 @@
 		return FALSE
-	. = COMPONENT_NO_ATTACK
+	if(!istype(O, /obj/item/slimecross/reproductive))
+		. = COMPONENT_NO_ATTACK
```
I dont like that this is done on all storage things for everything for a reproductive override. Can you find where preattack_intercept is called from and override it on the slimecross instead?

PowerfulBacon replied:
(override pre_attack on the reproductive extract before ..())

PowerfulBacon reviewed (changes requested):
^

MNarath1 committed:
lets instead override the proc for biobags

this adds the same check to the override pre_attack proc for biobags so it will return TRUE if A is a reproductive crossbreed.

MNarath1 commented:
Well i had to override it at the pre_attack proc for biobag but yeah this works

francinum commented:
This should be far more generic than it is. Arguably I'm not sure it's a good idea to allow this *in general*

MNarath1 commented:
> This should be far more generic than it is. Arguably I'm not sure it's a good idea to allow this _in general_

Please specify you refering to the crossbreed feeding with the biobag at its core or the proc override? 
First one this is how it should work so yeah.
For the second this proc is the one where it comes down now why the behavior changed exactly i am not sure honestly only that it all comes down to this proc and if said proc returns TRUE or FALSE now i was thinking of just overriding this completly and leaving the check out trough that probably would mess up more stuff then help honestly so i figured just including a single check might be better here.

MNarath1 commented:
Calling the parent would also do no good cause that would just call return immidiatly so thats the reason why i have this copy pasta.

francinum commented:
Items in storage are inherently somewhat resistant to manipulation due to the fact that we cannot easily predict their side effects and thus cannot account for actions that could put the item into an illegal state (overloading a storage container, for instance)

MNarath1 commented:
> Items in storage are inherently somewhat resistant to manipulation due to the fact that we cannot easily predict their side effects and thus cannot account for actions that could put the item into an illegal state (overloading a storage container, for instance)

Well i cannot completly gurantee this won't happen but from what i see it i simply restore old behavior trough not at the topmost point this if check was false at least for cases like reproductive extracts thats for sure now this suddenly changed so i simply added another check to see if it is a reproductive extract the rest should work exactly like before so i find it unlikely something will suddenly break.

PowerfulBacon commented:
The reason this used to work I presume is because the bio bag had click_gather which prevents preattack from blocking calls.

Archanial commented on code/game/objects/items/storage/bags.dm:
```diff
@@ -391,6 +391,11 @@
 	STR.can_hold = typecacheof(list(/obj/item/slime_extract, /obj/item/reagent_containers/syringe, /obj/item/reagent_containers/dropper, /obj/item/reagent_containers/glass/beaker, /obj/item/reagent_containers/glass/bottle, /obj/item/reagent_containers/blood, /obj/item/reagent_containers/hypospray/medipen, /obj/item/reagent_containers/food/snacks/deadmouse, /obj/item/reagent_containers/food/snacks/mo ...
 
+/obj/item/storage/bag/bio/pre_attack(atom/A, mob/living/user, params)
+	if(SEND_SIGNAL(src, COMSIG_ITEM_PRE_ATTACK, A, user, params) & COMPONENT_NO_ATTACK & !istype(A, /obj/item/slimecross/reproductive))
```
This isn't right, you don't use single & for bool checks

MNarath1 replied:
Right no shortcircuit behavior

MNarath1 committed:
shortcircuitgood

MNarath1 commented:
> The reason this used to work I presume is because the bio bag had click_gather which prevents preattack from blocking calls.

For that click_gather would have to be FALSE cause the check is  || !click_gather|| and then the proc would return FALSE that would indeed be consistant but i cannot see where this was actually changed in the last few months at all.

Archanial commented on code/game/objects/items/storage/bags.dm:
```diff
@@ -391,6 +391,11 @@
 	STR.can_hold = typecacheof(list(/obj/item/slime_extract, /obj/item/reagent_containers/syringe, /obj/item/reagent_containers/dropper, /obj/item/reagent_containers/glass/beaker, /obj/item/reagent_containers/glass/bottle, /obj/item/reagent_containers/blood, /obj/item/reagent_containers/hypospray/medipen, /obj/item/reagent_containers/food/snacks/deadmouse, /obj/item/reagent_containers/food/snacks/mo ...
 
+/obj/item/storage/bag/bio/pre_attack(atom/A, mob/living/user, params)
+	if(SEND_SIGNAL(src, COMSIG_ITEM_PRE_ATTACK, A, user, params) & COMPONENT_NO_ATTACK && !istype(A, /obj/item/slimecross/reproductive))
```
```suggestion
	if((SEND_SIGNAL(src, COMSIG_ITEM_PRE_ATTACK, A, user, params) & COMPONENT_NO_ATTACK) && !istype(A, /obj/item/slimecross/reproductive))
```

MNarath1 committed:
Update code/game/objects/items/storage/bags.dm

Co-authored-by: Archanial <37456678+Archanial@users.noreply.github.com>

francinum commented:
Snowflake. This change was made for a reason.

No.

MNarath1 commented:
> Snowflake. This change was made for a reason.
> 
> No.

wait this was an intended change???
There is no mention of it and the proc for it is there too also i found this PR on tg https://github.com/tgstation/tgstation/pull/60345 so it seems that it is indeed a bug trough might just port this now as it at least will not require snowflake child proc and also refractors the whole thing a bit

Archanial commented:
Because it's a bug lol
